python/pyarrow/pandas_compat.py

```diff
@@ -361,23 +351,78 @@ def _get_columns_to_convert(df, schema, preserve_index, columns):
         convert_types.append(type)
         column_names.append(name)
 
-    for i, column in enumerate(index_columns):
-        columns_to_convert.append(column)
-        convert_types.append(None)
-        name = _index_level_name(column, i, column_names)
-        index_column_names.append(name)
+    index_descriptors = []
+    index_column_names = []
+    for i, index_level in enumerate(index_levels):
+        name = _index_level_name(index_level, i, column_names)
+        if isinstance(index_level, pd.RangeIndex):
+            descr = _get_range_index_descriptor(index_level)
+        else:
+            columns_to_convert.append(index_level)
+            convert_types.append(None)
+            descr = {
+                'kind': 'serialized',
+                'field_name': name
+            }
+            index_column_names.append(name)
+        index_descriptors.append(descr)
+
+    all_names = column_names + index_column_names
+
+    # all_names : all of the columns in the resulting table including the data
+    # columns and serialized index columns
+    # column_names : the names of the data columns
+    # index_descriptors : descriptions of each index to be used for
+    # reconstruction
+    # index_levels : the extracted index level values
+    # columns_to_convert : assembled raw data (both data columns and indexes)
+    # to be converted to Arrow format
+    # columns_types : specified column types to use for coercion / casting
+    # during serialization, if a Schema was provided
+    return (all_names, column_names, index_descriptors, index_levels,
+            columns_to_convert, convert_types)
 
-    names = column_names + index_column_names
 
-    return (names, column_names, index_columns, index_column_names,
-            columns_to_convert, convert_types)
+def _get_range_index_descriptor(level):
+    # TODO(wesm): Why are these non-public and is there a more public way to
+    # get them?
+    return {
+        'kind': 'range',
+        'name': level.name,
+        'start': level._start,
+        'stop': level._stop,
+        'step': level._step
+    }
```

Comment:
@jreback @jorisvandenbossche is there a better way to get the start/stop/step for RangeIndex?

Reply:
As far as I know, we don't expose those publicly, so the private ones you are using here is the best pandas has to offer. 
They also have been stable over the years, so I don't think it is a problem to use them. I don't really recall the history of not exposing them (maybe because we don't really want end-users to rely too much on the fact it is special, but just see it as a memory optimized integer index).